### PR TITLE
Create separate cmd flag for platform none

### PR DESCRIPTION
**Pull request by alknopfler**
This pull request contains:

- Command create cluster with new flag "platform none" to avoid creating all the aws infra when you want to use platform none. 
- Basically hosted clusters will be deployed without any nodepool (by now) in the guest ocp cluster. 
- Using nodeport for kubeapi-server and konnectivity and getting dinamically the ip of master_0 guest node as annotation (to fit with SNO and multinode clusters)
- Don't need the aws credential because the control plane will be deployed in the guest ocp cluster.

Example of command to use with this PR:
- install the hypershift compiled previously: `[root@qct-d14u03 bin]# ./hypershift install --hypershift-image quay.io/amorgant/hypershift:v0.0.37`
- CLI command render manifests with the platform flag none and nodepool -2 to avoid using nodepool (by now): `./hypershift create cluster --pull-secret ../../pullsecret.json --ssh-key ~/.ssh/id_rsa.pub --base-domain alklabs.com --namespace clusters2 --node-pool-replicas -2 --release-image quay.io/openshift-release-dev/ocp-release:4.8.6-x86_64 --platform none --render`
- CLI command: `./hypershift create cluster --pull-secret ../../pullsecret.json --ssh-key ~/.ssh/id_rsa.pub --base-domain alklabs.com --namespace clusters2 --node-pool-replicas -2 --release-image quay.io/openshift-release-dev/ocp-release:4.8.6-x86_64 --platform none`
 
Results:  `[root@qct-d14u03 bin]#  oc get hostedclusters -n clusters2
NAME      VERSION   KUBECONFIG                 PROGRESS   AVAILABLE   REASON
example             example-admin-kubeconfig   Partial    True        HostedClusterAsExpected
[root@qct-d14u03 bin]#`

### CODE REVIEW

File: api/fixtures/example.go
Comment (sjenning) [inline api/fixtures/example.go:261]: Could we take this as a `none` only CLI flag rather than a bespoke annotation

File: api/fixtures/example.go
Comment (sjenning) [inline api/fixtures/example.go:282]: After https://github.com/openshift/hypershift/pull/458, this can be `Route`

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go
Comment (sjenning) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:835]: Why is this removed? Rebase issue or on purpose?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go
Comment (sjenning) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:539]: I am looking to remove this in https://github.com/openshift/hypershift/pull/462

File: cmd/cluster/create.go
Comment (sjenning) [inline cmd/cluster/create.go:79]: Why do we want to set a default for this?  `example.com` doesn't seem like it would ever work outside a local test environment.

File: cmd/cluster/create.go
Comment (sjenning) [inline cmd/cluster/create.go:295]: It seems that the only need for this is to look up the `master_0.Address`.  Could we take that on the CLI and have no `infra` stage for platform `None`?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go
Comment (sjenning) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:55]: This is already imported on line 54.  Lets not find/replace the name of this import i.e. use `client` not `ctrlclient`.  This was recently changed in https://github.com/openshift/hypershift/pull/419 so maybe this is just a rebase issue.

**Review by sjenning — CHANGES_REQUESTED** [review]

File: cmd/cluster/create.go
Comment (alknopfler) [reply cmd/cluster/create.go:295]: By now, I think you're right, but the idea is prepare the function to deploy infra for baremetal when CAPI will be implemented.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go
Comment (alknopfler) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:55]: you're right

File: api/fixtures/example.go
Comment (alknopfler) [reply api/fixtures/example.go:282]: ok makes sense

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go
Comment (alknopfler) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:835]: rebase issue

File: cmd/cluster/create.go
Comment (alknopfler) [reply cmd/cluster/create.go:294]: My idea is basically prepare for the next contribution to CAPI for baremetal. This is why I prefer create the new function to allocate the infra creation.

File: cmd/cluster/create.go
Comment (alknopfler) [reply cmd/cluster/create.go:79]: agree! It was an example but much better force to flag

File: api/fixtures/example.go
Comment (alknopfler) [reply api/fixtures/example.go:244]: Sorry, but I don't understand it :( My idea is just to use the k8s client to get the master0.Address instead of asking/retrieving the information with the flag. With master_0 you fit for SNO/Multinode guest cluster.

Comment (sjenning) [reply api/fixtures/example.go:243]: Yeah, my comment wasn't clear.

I was thinking about taking this as a flag on `create cluster` like `--external-api-server-address` that is only used if `--platform=none` is also set.

Comment (alknopfler) [reply api/fixtures/example.go:243]: It could be... but why do you need it? I mean, if you have the information for the master address for guest cluster, why should be added by an extra flag? Remember that this information is only for a NodePort Address until Route will be developed/available. That's why I decided to use the master0 Ip address until 'Route' will be available (the same to Konnectivity as you did in the other PR).

Comment (ironcladlou) [reply api/fixtures/example.go:243]: Then again... do we? I thought we originally had discussed constraining the scope of the CLI to the bare minimum necessary to support the "kick the tires" or demonstrative use case with AWS (and to provide some baseline smoke test for CI).

For everything else, I thought we had decided users should be constructing their manifests using their own more sophisticated tooling.

I'd like to know how we can draw a more clear line with regards to the intended use cases for the CLI so that we can define and avoid scope creep.

Comment (sjenning) [reply api/fixtures/example.go:243]: @ironcladlou My issue is that one can't easily get a HostedCluster/NodePool spec out of the CLI without side effects.

Right now, I run the CLI with `--render > cluster.yaml` which requires AWS creds and creates a ton of resources in AWS.  Then I immediately run the `destroy cluster` command to delete them and make the manual modification to `cluster.yaml` for platform `None` before doing an `oc create -f cluster.yaml`.

Our current unstable API means there are no such higher level tools and we are left to explain repetitively how to convert an platform `AWS` HostedCluster/NodePool to a platform `None`.

I do think the CLI should have some way to get a basic HostedCluster/NodePool manifest out with no side effects.

Comment (ironcladlou) [reply api/fixtures/example.go:243]: Makes sense... I like the idea of making these quality of life improvements. I do wonder in a bigger picture way where else we could be focusing on UX (CLI commands, templates in $tempate_system, static examples, API documentation, example/guide type documentation) generally, but we can take that conversation elsewhere.

Thanks for working on this!

Comment (alknopfler) [reply api/fixtures/example.go]: > Yeah, my comment wasn't clear.
> 
> I was thinking about taking this as a flag on `create cluster` like `--external-api-server-address` that is only used if `--platform=none` is also set.

changed.

File: cmd/cluster/create.go
Comment (alknopfler) [reply cmd/cluster/create.go:295]: Changed with options from flag instead of looking for the address

File: api/fixtures/example.go
Comment (enxebre) [inline api/fixtures/example.go:188]: platform names should live into a constant

File: api/fixtures/example.go
Comment (enxebre) [inline api/fixtures/example.go:236]: platform names should live into a constant.

File: cmd/cluster/create.go
Comment (enxebre) [inline cmd/cluster/create.go:91]: platform names should live into a constant.

File: cmd/cluster/create.go
Comment (enxebre) [inline cmd/cluster/create.go:112]: It does not necessarily need to be for baremetal. Let's avoid the term baremetal for now until we have a clear roadmap for all the uses cases and stories in baremetal environments.

File: cmd/cluster/create.go
Comment (enxebre) [inline cmd/cluster/create.go:116]: please remove commented code

File: cmd/cluster/create.go
Comment (enxebre) [inline cmd/cluster/create.go:169]: let's use constants for platforms names.

File: cmd/cluster/create.go
Comment (enxebre) [inline cmd/cluster/create.go:171]: please remove the term baremetal (https://github.com/openshift/hypershift/pull/452/files#r705149913) and let's use constants for platform names i.e None.

File: cmd/cluster/create.go
Comment (enxebre) [inline cmd/cluster/create.go:179]: why is this needed?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:414]: the previous message was concise, why change it?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:420]: the previous message was concise, why change it?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:1687]: nit: let's be consistent with the rest of the code base and use block comment style. Inlining here does to seems to add value.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:1708]: should probably not create a cluster at all. It will never reconcile successfully without an infra CR.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:1686]: why is this `var err error` needed?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go
Comment (enxebre) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:1708]:
```
if infra == nil {
 return early
}

Otherwise instantiate Cluster
```

File: cmd/cluster/create.go
Comment (alknopfler) [reply cmd/cluster/create.go:179]: error initialization for "infra, err = opt.CreateInfra(ctx)"

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go
Comment (alknopfler) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:414]: my bad with a replace

**Comment by alknopfler** [comment]
@enxebre changed

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go
Comment (enxebre) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:1687]: sorry I was missreading before. I think the most readable way would be:

```
	cluster.Spec = capiv1.ClusterSpec{
		ControlPlaneEndpoint: capiv1.APIEndpoint{},
		ControlPlaneRef: &corev1.ObjectReference{
			APIVersion: "hypershift.openshift.io/v1alpha1",
			Kind:       "HostedControlPlane",
			Namespace:  hcp.Namespace,
			Name:       hcp.Name,
		},
		InfrastructureRef: &corev1.ObjectReference{},
	}

if infra != nil {
 cluster.Spec.InfrastructureRef= &corev1.ObjectReference{
			APIVersion: gvk.GroupVersion().String(),
			Kind:       gvk.Kind,
			Namespace:  infraCR.GetNamespace(),
			Name:       infraCR.GetName(),
		},
}

```

We should pobably not create the CR at all, i.e don't call
```
	capiCluster := controlplaneoperator.CAPICluster(controlPlaneNamespace.Name, hcluster.Spec.InfraID)
	_, err = controllerutil.CreateOrUpdate(ctx, r.Client, capiCluster, func() error {
		return reconcileCAPICluster(capiCluster, hcluster, hcp, infraCR)
	})
```
We can discuss and do that in a different PR though

Comment (alknopfler) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:1687]: sorry but I don't know what you want to say. Let's talk tomorrow just to discuss ok?

**Comment by enxebre** [comment]
/test e2e-aws-pooled
/lgtm
/hold
for @ironcladlou and @sjenning to merge

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go
Comment (ironcladlou) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:428]: Why is the error handling being deleted here?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go
Comment (ironcladlou) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:456]: Why is the error handling being deleted here?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go
Comment (ironcladlou) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:481]: Why is the error handling being deleted here?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go
Comment (ironcladlou) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:879]: Why is the error handling being deleted here?

**Review by ironcladlou — CHANGES_REQUESTED** [review]

File: cmd/cluster/create.go
Comment (sjenning) [inline cmd/cluster/create.go:303]: don't need else after return

File: cmd/cluster/create.go
Comment (sjenning) [reply cmd/cluster/create.go:183]: i also don't see why this is needed

File: cmd/cluster/create.go
Comment (ironcladlou) [inline cmd/cluster/create.go:91]: Why was this changed? The defaulting is [here](https://github.com/openshift/hypershift/blob/745a6c3a4b6f098281e56623fbfdc7f89784e171/cmd/cluster/create.go#L204) and setting a default on the flag like this makes the CLI present a lie to the user (as the printed value is random per invocation; running `--help` would show one value, but by actually running it again would use a different random value). What problem was changing this solving?

File: api/fixtures/example.go
Comment (ironcladlou) [inline api/fixtures/example.go:14]: Platforms [already have constant types](https://github.com/openshift/hypershift/blob/745a6c3a4b6f098281e56623fbfdc7f89784e171/api/v1alpha1/hostedcluster_types.go#L233-L244) you should use

File: cmd/cluster/create.go
Comment (ironcladlou) [inline cmd/cluster/create.go:25]: Platforms [already have constant types](https://github.com/openshift/hypershift/blob/745a6c3a4b6f098281e56623fbfdc7f89784e171/api/v1alpha1/hostedcluster_types.go#L233-L244) you should use

File: cmd/cluster/create.go
Comment (ironcladlou) [inline cmd/cluster/create.go:303]: No need for the `else` given the `if` returns

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go
Comment (sjenning) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:428]: Just create a noop switch case for None rather than removing the error handling on the AWS path

File: hypershift-operator/controllers/nodepool/nodepool_controller.go
Comment (sjenning) [inline hypershift-operator/controllers/nodepool/nodepool_controller.go:547]: don't need "none" here

File: api/fixtures/example.go
Comment (sjenning) [inline api/fixtures/example.go:15]: let use the consts `hyperv1.AWSPlatform` and `hyperv1.NonePlatform`

**Comment by enxebre** [comment]
Can you please include the PR desc as the commit desc, i.e `git ci -m"Imperative mood commit subject" -m"Commit desc goes here."`

**Comment by alknopfler** [comment]
@enxebre , @ironcladlou , @sjenning refactor with the comments done, ready to review again. Tests e2e also passed

File: cmd/cluster/create.go
Comment (enxebre) [inline cmd/cluster/create.go:222]: wouldn't this whole thing need to within `if opts.Platform == string(hyperv1.AWSPlatform) {`

This lgtm overall. Either here or in a follow up since we are growing in platforms I think we should consider a more structure flow with may be specific functions per providers.

**Comment by alknopfler** [comment]
> @alknopfler are you still wanting to get this to a mergeable state?

Yes, everything passed.

**Comment by enxebre** [comment]
/hold
Thanks @alknopfler! this seems reasonable as to enabling platform None impl details. Before proceeding any further with this I want to make sure we set the basis in the code base to support a sustainable extensibility model for immediate upcoming providers e.g azure. We need to make sure to model and encapsulate provider agnostic/specific info in a way that is easy to add new providers and also should provide an easy and consistent UX for the different platforms commands/flags. I'll propose something then we can follow up with the platform none impl details added here.

**Comment by enxebre** [comment]
@alknopfler https://github.com/openshift/hypershift/pull/630 is to enable CLI to be extended for multiple providers in a sustainable way. After it gets merged let's incorporate this PR to ensure for simplicity that CAPI is consistent with platform None.

**Comment by sjenning** [comment]
This was done in https://github.com/openshift/hypershift/pull/630

**Pull request closed without merging**
